### ISSUE TITLE
#suggestion Infix operators grouping and "not match"

### ISSUE THREAD

**Issue by adri326** [ponylang/rfcs#139]
## Infix operators grouping

The "no operator precedence" rule has been quite new to me, but I am getting used to it.
What I would suggest, however, would be to soften the rule by introducing groups of infix operators: (infix) operators of the same group could be chained without needing to put parentheses around them, just like you would be able to do with only one operator:

```pony
a + b - c + d
a * b / c * d
```

The above examples would then be valid, and the unsuggared version would look like this:

```pony
a.add(b).sub(c).add(d)
a.mul(b).div(c).mul(d)
```

The groups would be:

* `+` (add) and `-` (sub)
* `*` (mul) and `/` (div)
* *Eventually*: `<<` (lshift) and `>>` (rshift) 

Now, to why I am doing this proposal: on a mathematical standpoint, doing `a + b - c` and `a - c + b` is essentially the same thing: you have at the end added `a` and `b` and have substracted at some point `c`; all operations will be done from left to right, without any concept of precedence being at play.
When writing code, one could think to: "add, then substract, then add" his values, as in a big sumation, and would so write: `a + b - c + d`. However, pony's rule will force him to put parentheses around his code: `((a + b) - c) + d` or `(a + b + d) - c`, essentially changing what they were thinking first ("first add, then substract to the result, then add to the result" or "add the three then substract to the result").
Surely, they could've written `a + b + -c + d`, which, in a mathematical way, is also correct, but this wouldn't be possible with the other group(s) of operators.

Implementing this would not break any code already done, as it only allows more to be done by softening the rules.
Theses rules would look like this:

```
Repeated use of a single operator, or of operators of the same groups, is fine:

`1 + 2 + 3`

<mention operator groups>

`1 + 2 - 3`
```

That was it for the infix operators grouping.

## "not match"

This is just a simple edge case: doing `not match ... end` results in an error:

```
syntax error: expected expression after not
```

The same error is raised when trying to do `not if`, it is easily fixed when putting parentheses around the `match` or the `if` expression. It bugs me a little as it is said that `match`es and `if`s are considered as expressions.
Here is an example:

https://playground.ponylang.io/?gist=2652feb86d95f77db66c4b1589358beb

Let me know if I should move this issue to another github issue.

Thanks for reading up to here, if you did, and have a nice day!

**Comment by EpicEric** [comment]
I have some points against your first suggestion, which I will express below. As for your second suggestion, it makes perfect sense to me.

TL;DR: There are several cases where infix operators grouping can lead to bugs or confusion.

---

Grouping different infix operators would implicitly assume that they have the same order of precedence. This would create some arbitration as to whether operators have the same precedence or not, [and](https://tutorial.ponylang.io/expressions/infix-ops.html#precedence):

>The problem with this is that the programmer has to remember the order and people aren’t very good at things like that. Most people will remember to do multiplication before addition, but what about left bit shifting versus bitwise and? Sometimes people misremember (or guess wrong) and that leads to bugs. Worse, those bugs are often very hard to spot.

This also applies to "+" and "-", which may not have the same precedence to another person in another culture, or even with a different math system. Say that I want to create a module that uses these operators, where they aren't directly reversible operations (let's say, "OR" and "NOR" on U32 instead of booleans, respectively). This would be a case where `((a + b) - c) + d` and `(a + b + d) - c` mean two different things, and infix operator grouping would silently cause bugs. [Exhibit A](https://playground.ponylang.io/?gist=deacd37aca1a8a4894a96794f53369a2).

(Note that this issue doesn't happen when there's only one type of infix operator, because the structure of the code itself already implies left-to-right ordering, and an operator cannot have a different precedence than itself.)

Even in our regular math system, they are not the same thing. `(a + b + d) - c` could have rounding errors with big floats and give a different result than `((a + b) - c) + d`, due to an intrinsic limitation of numerical representations in CPUs. [Exhibit B](https://playground.ponylang.io/?gist=a8bd5485e9faeea14247925505075d17).

For these reasons, I agree with the language's design decision to keep any kind of operator precedence out of the syntax, for the sake of avoiding potentially unmaintainable code.

**Comment by adri326** [comment]
Your point is right: I was thinking in terms of good ol' mathematical plus and minuses, when one could actually have the `.add` method on their class do a multiplication and the `.mul` method do an addition, making my argument pointless.
I wasn't thinking about the fact that the infix operators can have different meanings depending on their uses in the user code, these different meanings making any grouping irrelevant.

Thank you for pointing me this out!

**Comment by SeanTAllen** [comment]
Please split these into two different issues as they are two different issues. Also please note, that these are issues which means you would be looking for conversation and possibly someone to write an RFC for either issue. Nothing would be done one way or another without an actual RFC.

**Comment by SeanTAllen** [comment]
@EpicEric your point...

```
Even in our regular math system, they are not the same thing. (a + b + d) - c could have rounding errors with big floats and give a different result than ((a + b) - c) + d, due to an intrinsic limitation of numerical representations in CPUs. Exhibit B.
```

Is an interesting one. If you are up for it, I think it would be interesting to add the general discussion to the operator precedence section of the tutorial. Would you be up for PRing that?

**Comment by adri326** [comment]
Opened a new issue, I guess this issue can be closed
